warehouse/query-core/src/main/java/datawave/next/DocIdIteratorVisitor.java

```diff
@@ -0,0 +1,521 @@
+        if (isDocRange) {
+            iterator.withSuffix(getSuffix());
+        }
+
+        // check to see if this scan exists within the bounds of another scan
+        if (data instanceof ScanResult) {
+            ScanResult scanResult = (ScanResult) data;
+            iterator.withMinMax(scanResult.getMin(), scanResult.getMax());
+        }
+
+        // if scan results exist then this scan is allowed to timeout
+        boolean checkForTimeout = data instanceof ScanResult;
+        long scanStart = System.currentTimeMillis();
+        long elapsedScanTime;
+
+        int count = 0;
+        ScanResult result = new ScanResult();
+        while (iterator.hasNext()) {
+            count++;
+            result.addKey(iterator.next());
+            if (checkForTimeout && count % resultInterval == 0) {
+                elapsedScanTime = System.currentTimeMillis() - scanStart;
+                if (elapsedScanTime >= maxScanTimeMillis) {
+                    result.setTimeout(true);
+                    log.warn("term: [{}] founds {} hits before hitting timeout threshold: {}", iterator.getNode(), result.getResults().size(),
+                                    maxScanTimeMillis);
+                    break;
+                }
+            }
+        }
+
+        elapsedScanTime = System.currentTimeMillis() - scanStart;
+        stats.merge(iterator.getStats());
+
+        if (log.isDebugEnabled()) {
+            log.debug("term: [{}] found {} hits in {} ms", iterator.getNode(), result.getResults().size(), elapsedScanTime);
+        }
+        return result;
+    }
+
+    private String getSuffix() {
+        return range.getStartKey().getColumnFamily().toString();
+    }
+
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        return node.jjtGetChild(0).jjtAccept(this, data);
+    }
+
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        return node.jjtGetChild(0).jjtAccept(this, data);
+    }
+
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        return node.jjtGetChild(0).jjtAccept(this, data);
+    }
+
+    @Override
+    public Object visit(ASTNRNode node, Object data) {
+        return null;
+    }
+
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        return null;
+    }
+
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        return null;
+    }
+
+    @Override
+    public Object visit(ASTMethodNode node, Object data) {
+        return null;
+    }
+
+    public DocumentIteratorStats getStats() {
+        return stats;
+    }
+
+    public void setMaxScanTimeMillis(long maxScanTimeMillis) {
+        this.maxScanTimeMillis = maxScanTimeMillis;
+    }
+
+    /**
+     * This class allows us to retain knowledge of the min and max keys from a scan without requiring an expensive sorted set. Future scan ranges are restricted
+     * by the min and max
+     */
```

Comment:
when doing an intersection?

Reply:
This is what allows future scans to have a more restrictive start and end key, when applicable.